src/Symfony/Component/ArgumentResolver/ArgumentMetadata/ArgumentMetadata.php

```diff
@@ -0,0 +1,146 @@
+<?php
+
+/*
+ * This file is part of the Symfony package.
+ *
+ * (c) Fabien Potencier <fabien@symfony.com>
+ *
+ * For the full copyright and license information, please view the LICENSE
+ * file that was distributed with this source code.
+ */
+
+namespace Symfony\Component\ArgumentResolver\ArgumentMetadata;
+
+/**
+ * Responsible for storing metadata of an argument.
+ *
+ * @author Iltar van der Berg <kjarli@gmail.com>
+ *
+ * @final
+ */
+class ArgumentMetadata
+{
+    public const IS_INSTANCEOF = 2;
+
+    /**
+     * @param object[] $attributes
+     */
+    public function __construct(
+        private string $name,
+        private ?string $type,
+        private bool $isVariadic,
+        private bool $hasDefaultValue,
+        private mixed $defaultValue,
+        private bool $isNullable = false,
+        private array $attributes = [],
+        private string $callableName = 'n/a',
+    ) {
+        $this->isNullable = $isNullable || null === $type || ($hasDefaultValue && null === $defaultValue);
+    }
+
+    /**
+     * Returns the name as given in PHP, $foo would yield "foo".
+     */
+    public function getName(): string
+    {
+        return $this->name;
+    }
+
+    /**
+     * Returns the type of the argument.
+     *
+     * The type is the PHP class in 5.5+ and additionally the basic type in PHP 7.0+.
+     */
+    public function getType(): ?string
+    {
+        return $this->type;
+    }
+
+    /**
+     * Returns whether the argument is defined as "...$variadic".
+     */
+    public function isVariadic(): bool
+    {
+        return $this->isVariadic;
+    }
+
+    /**
+     * Returns whether the argument has a default value.
+     *
+     * Implies whether an argument is optional.
+     */
+    public function hasDefaultValue(): bool
+    {
+        return $this->hasDefaultValue;
+    }
+
+    /**
+     * Returns whether the argument accepts null values.
+     */
+    public function isNullable(): bool
+    {
+        return $this->isNullable;
+    }
+
+    /**
+     * Returns the default value of the argument.
+     *
+     * @throws \LogicException if no default value is present; {@see self::hasDefaultValue()}
+     */
+    public function getDefaultValue(): mixed
+    {
+        if (!$this->hasDefaultValue) {
+            throw new \LogicException(\sprintf('Argument $%s does not have a default value. Use "%s::hasDefaultValue()" to avoid this exception.', $this->name, __CLASS__));
+        }
+
+        return $this->defaultValue;
+    }
+
+    /**
+     * @param class-string          $name
+     * @param self::IS_INSTANCEOF|0 $flags
+     *
+     * @return array<object>
+     */
+    public function getAttributes(?string $name = null, int $flags = 0): array
+    {
+        if (!$name) {
+            return $this->attributes;
+        }
+
+        return $this->getAttributesOfType($name, $flags);
+    }
```

Comment:
Either we make the `getAttributesOfType` method private, or we remove the arguments of this one. Having 2 methods that get the same arguments and return the same result is not ideal.

```suggestion
    /**
     * @return array<object>
     */
    public function getAttributes(): array
    {
        return $this->attributes;
    }
```

The backward-compatibility layer can be kept and deprecated in the deprecated child class.